File: src/memory/builder/roadmap_position.py
"""Resolve Builder roadmap position from roadmap files."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from memory.builder.roadmap_grammar import (
    HEADING_RE as _HEADING_RE,
)
from memory.builder.roadmap_grammar import (
    STATUS_RE as _STATUS_RE,
)
from memory.builder.roadmap_grammar import (
    is_legacy_path as _is_legacy_path,
)

_ACTIVE_STATUS_MARKERS = ("🟢 Active", "Active")


@dataclass(frozen=True)
class RoadmapPosition:
    code: str
    title: str
    status: str
    path: str


def resolve_roadmap_position(project_path: Path | None) -> RoadmapPosition | None:
    """Return the first active roadmap file under a project, if any."""
    if project_path is None:
        return None
    root = project_path.expanduser().resolve()
    roadmap_root = root / "docs" / "project" / "roadmap"
    if not roadmap_root.is_dir():
        return None

    candidates: list[RoadmapPosition] = []
    for path in sorted(roadmap_root.rglob("index.md")):
        if _is_legacy_path(path, roadmap_root):
            continue
        position = _position_from_file(root, path)
        if position is not None:
            candidates.append(position)
    return candidates[0] if candidates else None


def _position_from_file(root: Path, path: Path) -> RoadmapPosition | None:
    try:
        content = path.read_text(encoding="utf-8")
    except OSError:
        return None
    status_match = _STATUS_RE.search(content)
    if not status_match:
        return None
    status = status_match.group("status").strip()
    if not any(marker in status for marker in _ACTIVE_STATUS_MARKERS):
        return None
    heading_match = _HEADING_RE.search(content)
    if not heading_match:
        return None
    return RoadmapPosition(
        code=heading_match.group("code").strip(),
        title=heading_match.group("title").strip(),
        status=status,
        path=path.resolve().relative_to(root).as_posix(),
    )

File: ts/src/builder/roadmapPosition.ts
// CV22.DS7.US8 plateau 1 — the active roadmap position for the resume surface.
//
// Port of `src/memory/builder/roadmap_position.py`. `build load` renders this in
// the `roadmap position` row of `■ BUILDER RESUME`.
//
// It returns the FIRST active package in scan order, so it inherits the
// component-wise path ordering from `roadmapScan` — the same reason a string
// sort would change what the Navigator sees.
//
// `_ACTIVE_STATUS_MARKERS` is `("🟢 Active", "Active")`, which is a substring
// test where the second entry subsumes the first: any status containing
// "Active" qualifies, including "Inactive". Reproduced rather than tightened.

import { pyStrip } from "#util/pythonText.ts";
import { roadmapPaths } from "./pullCandidates.ts";
import { matchHeading, matchStatus } from "./roadmapGrammar.ts";
import { readRoadmapFile, scanRoadmapIndexFiles } from "./roadmapScan.ts";

const ACTIVE_STATUS_MARKERS = ["🟢 Active", "Active"] as const;

export interface RoadmapPosition {
  code: string;
  title: string;
  status: string;
  /** POSIX path relative to the project root, as Python's `.as_posix()`. */
  path: string;
}

/** Python `resolve_roadmap_position`: the first active roadmap file, or `null`. */
export function resolveRoadmapPosition(projectRoot: string | null): RoadmapPosition | null {
  if (projectRoot === null) return null;
  const { roadmapRoot } = roadmapPaths(projectRoot);
  for (const file of scanRoadmapIndexFiles(roadmapRoot)) {
    const content = readRoadmapFile(file.absolutePath);
    if (content === null) continue;
    const status = matchStatus(content);
    if (status === null) continue;
    const statusText = pyStrip(status);
    if (!ACTIVE_STATUS_MARKERS.some((marker) => statusText.includes(marker))) continue;
    const heading = matchHeading(content);
    if (!heading) continue;
    return {
      code: pyStrip(heading.code),
      title: pyStrip(heading.title),
      status: statusText,
      path: `docs/project/roadmap/${file.relativePath}`,
    };
  }
  return null;
}
